### PR TITLE
KAFKA-4225: Replication Quotas: Control Leader & Follower Limit Separately

### PR DESCRIPTION
**Pull request by benstopford**

### CODE REVIEW

File: core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala [hunk 1]
```diff
@@ -316,21 +318,25 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, proposedAssignment: Map[TopicA
   }
 
   def maybeLimit(throttle: Long) {
+    // Limit the throttle on currently moving replicas. Note that this command can use used to alter the throttle,
+    // but it may not alter all limits originally set, if some of the brokers have completed their rebalance.
```
Comment (ijuma) [inline core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:322]: I think we should have this and other similar comments as scaladoc as they are relevant for people invoking them.

File: core/src/main/scala/kafka/log/LogConfig.scala [hunk 2]
```diff
@@ -124,8 +124,8 @@ object LogConfig {
   val MessageFormatVersionProp = "message.format.version"
   val MessageTimestampTypeProp = "message.timestamp.type"
   val MessageTimestampDifferenceMaxMsProp = "message.timestamp.difference.max.ms"
-  val LeaderThrottledReplicasListProp = "quota.leader.replication.throttled.replicas"
-  val FollowerThrottledReplicasListProp = "quota.follower.replication.throttled.replicas"
+  val LeaderThrottledReplicasListProp = "leader.replication.throttled.replicas"
+  val FollowerThrottledReplicasListProp = "follower.replication.throttled.replicas"
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogConfig.scala:128]: What's the motivation for this change, make the names shorter? Oh, probably to make them consistent with the rate config names?

File: core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala [hunk 3]
```diff
@@ -24,7 +24,8 @@ object ReplicationQuotaUtils {
     TestUtils.waitUntilTrue(() => {
       val brokerReset = servers.forall { server =>
         val brokerConfig = AdminUtils.fetchEntityConfig(server.zkUtils, ConfigType.Broker, server.config.brokerId.toString)
-        !brokerConfig.contains(DynamicConfig.Broker.ThrottledReplicationRateLimitProp)
+        ! (brokerConfig.contains(DynamicConfig.Broker.ThrottledLeaderReplicationRateProp)
+          && brokerConfig.contains(DynamicConfig.Broker.ThrottledFollowerReplicationRateProp))
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala]: Did you mean to use `&&` here even though you use `||` for the topic case?

File: core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala [hunk 4]
```diff
@@ -38,8 +39,10 @@ object ReplicationQuotaUtils {
       //Check for limit in ZK
       val brokerConfigAvailable = servers.forall { server =>
         val configInZk = AdminUtils.fetchEntityConfig(server.zkUtils, ConfigType.Broker, server.config.brokerId.toString)
-        val zkThrottleRate = configInZk.getProperty(DynamicConfig.Broker.ThrottledReplicationRateLimitProp)
-        zkThrottleRate != null && expectedThrottleRate == zkThrottleRate.toLong
+        val zkLeaderRate = configInZk.getProperty(DynamicConfig.Broker.ThrottledLeaderReplicationRateProp)
+        val zkFollowerRate = configInZk.getProperty(DynamicConfig.Broker.ThrottledFollowerReplicationRateProp)
+        (zkLeaderRate != null && expectedThrottleRate == zkLeaderRate.toLong
+        && zkFollowerRate != null && expectedThrottleRate == zkFollowerRate.toLong)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala:45]: Nit: need one indent here.

File: core/src/main/scala/kafka/server/DynamicConfig.scala [hunk 5]
```diff
@@ -32,19 +33,25 @@ object DynamicConfig {
...
 
     //Defaults
-    val DefaultThrottledReplicationRateLimit = ReplicationQuotaManagerConfig.QuotaBytesPerSecondDefault
+    val DefaultThrottledReplicationRate = ReplicationQuotaManagerConfig.QuotaBytesPerSecondDefault
 
     //Documentation
-    val ThrottledReplicationRateLimitDoc = "A long representing the upper bound (bytes/sec) on replication traffic for replicas enumerated in the " +
-      s"property $ThrottledReplicationRateLimitProp. This property can be only set dynamically via the config command etc. The minimum value is 1 KB/s."
+    val ThrottledLeaderReplicationRateDoc = "A long representing the upper bound (bytes/sec) on replication traffic for leaders enumerated in the " +
+      s"property ${LogConfig.LeaderThrottledReplicasListProp} (for each topic). This property can be only set dynamically (via the kafka-configs.sh etc). It is suggested that the " +
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/DynamicConfig.scala]: Nit: would it read slightly better if we simply said `via kafka-configs.sh` (without the `the`)? If so, same for the follower doc.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Looks good overall. Left a few minor comments, but more importantly:
1. We need to make sure that the updated config names are good (would be good to run them by Jun)
2. Should we update the throttling tests to have one case where the rates between leader and follower are different?

File: core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala [hunk 6]
```diff
@@ -79,11 +79,13 @@ object ReassignPartitionsCommand extends Logging {
...
     //If all partitions have completed remove the throttle
-    if (reassignedPartitionsStatus.forall { case (topicPartition, status) => status == ReassignmentCompleted }) {
+    if (reassignedPartitionsStatus.forall { case (_, status) => status == ReassignmentCompleted }) {
       //Remove the throttle limit from all brokers in the cluster
+      //(as we no longer know which specific brokers were involved in the move)
       for (brokerId <- zkUtils.getAllBrokersInCluster().map(_.id)) {
         val configs = AdminUtils.fetchEntityConfig(zkUtils, ConfigType.Broker, brokerId.toString)
-        if (configs.remove(DynamicConfig.Broker.ThrottledReplicationRateLimitProp) != null){
+        if (configs.remove(DynamicConfig.Broker.ThrottledLeaderReplicationRateProp) != null
+          || configs.remove(DynamicConfig.Broker.ThrottledFollowerReplicationRateProp) != null){
```
Comment (junrao) [inline core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala]: Hmm, if the first test returns true, the code for removing follower rate won't be called?

File: core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala [hunk 7]
```diff
@@ -316,21 +318,25 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, proposedAssignment: Map[TopicA
...
     if (throttle >= 0) {
       val existingBrokers = existingAssignment().values.flatten.toSeq
       val proposedBrokers = proposedAssignment.values.flatten.toSeq
       val brokers = (existingBrokers ++ proposedBrokers).distinct
 
       for (id <- brokers) {
         val configs = AdminUtils.fetchEntityConfig(zkUtils, ConfigType.Broker, id.toString)
-        configs.put(DynamicConfig.Broker.ThrottledReplicationRateLimitProp, throttle.toString)
+        configs.put(DynamicConfig.Broker.ThrottledLeaderReplicationRateProp, throttle.toString)
+        configs.put(DynamicConfig.Broker.ThrottledFollowerReplicationRateProp, throttle.toString)
```
Comment (junrao) [inline core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:331]: It seems it makes more sense to take two throttle limits from the command line,  one for the leader and one for the follower?

File: core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala [hunk 8]
```diff
@@ -316,21 +318,25 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, proposedAssignment: Map[TopicA
...
         AdminUtils.changeBrokerConfig(zkUtils, Seq(id), configs)
       }
       println(s"The throttle limit was set to $throttle B/s")
     }
   }
 
   private[admin] def assignThrottledReplicas(allExisting: Map[TopicAndPartition, Seq[Int]], allProposed: Map[TopicAndPartition, Seq[Int]], admin: AdminUtilities = AdminUtils): Unit = {
+    //Set throttles to replicas that are moving. Note: this command should only be used when the assignment is initiated.
```
Comment (junrao) [inline core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala]: "this command" should be "this method"?

File: core/src/main/scala/kafka/server/DynamicConfig.scala (hunk 5 above)
Comment (junrao) [inline core/src/main/scala/kafka/server/DynamicConfig.scala]: This description will be shown when we run kafka-configs.sh. So, we probably don't need to mention kafka-configs.sh.

File: core/src/main/scala/kafka/server/DynamicConfig.scala [hunk 9]
```diff
@@ -32,19 +33,25 @@ object DynamicConfig {
...
+      s"limit be kept above the value of ${KafkaConfig.ReplicaFetchMaxBytesProp} which defaults to ${Defaults.ReplicaFetchMaxBytes}B/s"
```
Comment (junrao) [inline core/src/main/scala/kafka/server/DynamicConfig.scala]: Hmm, it's a bit weird to tie the quota limit to ReplicaFetchMaxBytesProp. The former is a rate, but the latter is a size. So, they are not directly comparable.

File: core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala (hunk 7 above)
Comment (benstopford) [reply core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:331]: So I'm not sure I agree. I see altering the throttle on leader-side or follower-side as a pretty advanced use case for people using the ConfigCommand. It'll be quite hard to document what affect different throttle values would have too. I suggest we keep it as is for now, but happy to discuss further if you feel strongly.

File: core/src/main/scala/kafka/server/DynamicConfig.scala (hunk 9 above)
Comment (benstopford) [reply core/src/main/scala/kafka/server/DynamicConfig.scala]: replaced with "limit be kept above 1MB/s for accurate behaviour." 
let me know if you'd like to tweak further.

File: core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala (hunk 7 above)
Comment (ijuma) [reply core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:331]: I think it makes sense to keep the command-line tool simple while having the flexibility built-in. Because convenience is important for command-line tools, even if we do add the ability to set the throttles separately in the command-line tool, it makes sense to make it possible to set both throttles via a single option as well. So, we can evolve this, if necessary, later.

**Comment by ijuma** [comment]
Thanks for the PR, LGTM.

I'm going to merge this as we are close to the code freeze and it's good to ensure that the configs, which are public API, are in their final state.

If @junrao still thinks that we should change the command-line tool to have two throttle rate parameters, we can do that in a follow-up as it's additive to what is done in this PR.

@benstopford, when you have a chance, can you please update the KIP document to include this change?

File: core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala (hunk 3 above)
Comment (benstopford) [reply core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala]: yes

File: core/src/main/scala/kafka/log/LogConfig.scala (hunk 2 above)
Comment (benstopford) [reply core/src/main/scala/kafka/log/LogConfig.scala:128]: yes. getting too long. not consistent
